### PR TITLE
feat(fees): impl supply/demand-adjusted fee

### PR DESCRIPTION
**Pull request by oetyng**
This PR adds an algorithm to calculate a fee with supply/demand adjusting properties - although indirect and therefore arguably sluggish - based on:
1. Network size (contributing to the deflationary character of SNT).
2. Storage used.
3. Requirements for space margin (1/3).
4. Requirement for as low fee as possible for as long as possible.

(Point 3. and 4. are achieved by the specific design of the required tokens curve.)

***

To do transfers on the network, clients are required to pay a number of tokens as a fee. The fees are the source of rewards for Elders in a payment-network, which is essential to the network security as it gives incentive to node operators to run a node and handle the requests, thus allowing for the network to grow.

**NB 1**: This is not a replacement for a true market, where node operators and clients discover the price based on operation costs + margins and service value + affordability to client.

**NB 2**: The fee paid to the Elder is not actually accessible until the Elder can fetch the `SpendProof` from
the data holders.

**NB 3**: With 1 fee per spend, the fee amount in the dbc is not accessible to the Elder until the required fee for
a spend has decreased below the amount in the dbc. In effect, with this design there is currently a lock-in
effect on Elder rewards which require the network to grow for the amount paid to them to be accessible.
***
1. ✔️ `Previous PR`: The query and handling of it on Elders for their reward keys and hard coded fee: #2184.
2. ✔️ `Previous PR`: The logic for including Elders to outputs: #2180.
3. ✔️ `Previous PR`: Including fee ciphers for Elders in spend: #2249.
4. ✔️ `Previous PR`: Elders confirming that spends contain outputs with sufficient amount, destined for them: #2246.
5. `This PR`: Use a fee calc algo with supply/demand adjusting properties.
6. `Next PR`: Use a mempool to increase supply/demand properties of the system and protect the network #2256.

Steps in the fee/reward feature:
- <s> 1. Including Elders in outputs (any value) when spending Dbcs.</s>
- <s> 2. Including fee ciphers for Elders so that they can confirm fee paid.</s>
- <s> 3. Elders confirming that spends contain outputs with sufficient amount, destined for them.</s>
- <s> 5. Implementing a commonly known deterministic transfer/store cost paid to nodes.</s> (`This PR`)
- <s> 5. Including supply/demand adjusting fee in outputs when spending Dbcs.</s> (`This PR`)
6. A mempool which orders incoming spends by fee amount, and a Client UI option to choose priority of their spend.
7. Elders fetching and storing their payment to a local wallet.

### CODE REVIEW

File: sn_interface/src/op_cost/mod.rs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use self_encryption::MAX_CHUNK_SIZE;
+use sn_dbc::Token;
+use tracing::debug;
+
+/// The conversion from token to raw value.
+const TOKEN_TO_RAW_CONVERSION: u64 = 1_000_000_000;
+/// The maximum supply of SNT, also the larges value that can be represented by a single `Token`.
```
Comment (maqi) [inline sn_interface/src/op_cost/mod.rs:15]: largest ?

File: sn_interface/src/op_cost/mod.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+/// 4. Requirement for as low fee as possible for as long as possible.
+/// (Point 3. and 4. are achieved by the specific design of the required tokens curve.)
+///
+/// It is assumed that a larger network means the token has a greater value, and therefore
+/// the required number of tokens per operation is also lower, following a curve as per above.
+/// The constants used may seem arbitrary but have been carefully chosen as to model the desired behaviour.
+pub fn required_tokens(
+    bytes: usize,
+    prefix_len: usize,
+    num_storage_nodes: u8,
```
Comment (maqi) [inline sn_interface/src/op_cost/mod.rs:38]: just for clarification, this `num_storage_nodes` actually means `all nodes across whole network` or just `nodes of our section`, or just `nodes supposed to store copies of this data` ?

File: sn_interface/src/op_cost/mod.rs [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    #[test]
+    fn calculates_required_tokens() {
+        let bytes = MAX_CHUNK_SIZE;
+        let prefix_len = 4;
+        let num_storage_nodes = 24;
+        let percent_filled = 3_f64 / num_storage_nodes as f64;
+        let required_tokens =
+            required_tokens(bytes, prefix_len, num_storage_nodes, percent_filled).as_nano();
+        println!("required_tokens: {required_tokens}");
+        assert_eq!(required_tokens, 15_300_364); // 0.01500364 tokens
```
Comment (maqi) [inline sn_interface/src/op_cost/mod.rs:78]: bit confused with this `0.01500364 tokens` , do you mean `0.01500364 coin` ?
As there are also comment of ` at least one token ...` indicating token is a very small measurement unit ?

File: sn_interface/src/op_cost/mod.rs (hunk 2 above)
Comment (oetyng) [reply sn_interface/src/op_cost/mod.rs:38]: It means nodes of our section. I will clarify.

**Review by maqi — APPROVED** [review]

**Review by davidrusu — DISMISSED** [review]

**Review by joshuef — APPROVED** [review]

**Review by RolandSherwin — APPROVED** [review]

**Pull request merged**